src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -1,44 +1,47 @@
 package mesosphere.marathon.tasks
 
-import org.apache.mesos.Protos
-
-import scala.util.{ Try, Random }
-import org.apache.mesos.Protos.{ Value, Offer }
+import mesosphere.marathon.state.{ AppDefinition, Container }
+import mesosphere.marathon.state.Container.Docker.PortMapping
 import mesosphere.mesos.protos
 import mesosphere.mesos.protos.{ RangesResource, Resource }
-import mesosphere.marathon.PortResourceException
-import mesosphere.marathon.state.AppDefinition
-import mesosphere.marathon.state.Container
-import mesosphere.marathon.state.Container.Docker.PortMapping
 import mesosphere.util.Logging
+import org.apache.mesos.Protos
+import org.apache.mesos.Protos.Offer
+
+import scala.annotation.tailrec
 import scala.collection.JavaConverters._
 import scala.collection.immutable.Seq
-
-object PortsMatcher {
-  case class PortRange(role: String, range: Protos.Value.Range) {
-    lazy val scalaRange: Range = range.getBegin.toInt to range.getEnd.toInt
-  }
-}
+import scala.util.Random
 
 /**
   * Utility class for checking if the ports resource in an offer matches the requirements of an app.
   */
-class PortsMatcher(app: AppDefinition, offer: Offer, acceptedResourceRoles: Set[String] = Set("*")) extends Logging {
-  import PortsMatcher._
-
-  private[this] lazy val offeredPortsResources: Seq[Protos.Resource] = offer.getResourcesList.asScala
-    .filter(resource => acceptedResourceRoles(resource.getRole))
-    .filter(_.getName == Resource.PORTS)
-    .to[Seq]
+class PortsMatcher(
+  app: AppDefinition,
+  offer: Offer,
+  acceptedResourceRoles: Set[String] = Set("*"),
+  random: Random = Random)
+    extends Logging {
 
-  private[this] lazy val offeredPortRanges: Seq[PortRange] = offeredPortsResources
-    .flatMap(resource => resource.getRanges.getRangeList.asScala.map(PortRange(resource.getRole, _)))
+  import PortsMatcher._
 
   /**
     * The resulting port matches which should be consumed from the offer. If no matching port ranges could
     * be generated from the offer, return `None`.
     */
-  lazy val portRanges: Option[Seq[RangesResource]] = {
+  lazy val portRanges: Option[Seq[RangesResource]] = portsWithRoles.map(PortWithRole.createPortsResources)
+
+  /**
+    * @return true if and only if the port requirements could be fulfilled by the given offer.
+    */
+  def matches: Boolean = portRanges.isDefined
+
+  /**
+    * @return the resulting assigned (host) ports.
+    */
+  def ports: Seq[Long] = portRanges.map(_.flatMap(_.ranges.flatMap(_.asScala()))).getOrElse(Nil)
```

Comment:
Probably the shortest expression possible, which is good, but also very hard to understand (especially with asScala I think of a java sequence).
Can you make that more readable?